### ISSUE TITLE
A plugin interface for adding video codecs

### ISSUE THREAD

**Issue by ericoporto** [adventuregamestudio/ags#1981]
**Describe the problem**
OGV works, and is ok for the video needs in a low resolution game. But it's compression is not very effective, which becomes noticeable in bigger resolutions. At the same time, video decoding libraries that are effective can be ten times the size of AGS itself, which make adding new codecs in the engine a hard problem.

**Suggested change**
Have some way in the plugin API to add new video codecs to AGS itself, in a way that the new format becomes playable by AGS, and the regular script functions for video are used. The idea is, if in the future some new video functionality is added in AGS, it should be able to leverage this API so it can access more codecs. It would probably be nice that, if the codec is not available it just degrades the experience by not playing the videos.

I don't have the API design tied down, but it should be so that the media/video internals of AGS could leverage such interface.

As an example, [here are builds of libvpx](https://github.com/ShiftMediaProject/libvpx), which would enable decoding of vp9 and vp8 `.webm` videos, you can see the dll files are around 20MB.

**Comment by ivan-mogilko** [comment]
Decoding video is a combination of decoding audio and decoding images, therefore I propose to start with these.
FFmpeg has a very flexible interface, although I do not suggest adding fully same, but it may be a good reference for the design.

The basic idea is this: a codec is something that converts from input data to output data. So in the most primitive case you have a function Convert with input and output buffers as arguments. But there's a problem: a codec may require certain number of data to even begin decoding, so it may ask you to keep passing more until it's ready to return something. On another hand, our engine not necessarily will be able to accept whole returned data (it may be inconvenient).
This brings us to extended different API, where we have separate PutInputData / GetOutputData methods. The user would put data while it can and until codec will tell that no more is necessary / possible to be put; and retrieve as much as wanted and as much as codec can give back.
This kind of interface would work virtually for anything (image, audio, video, any thing else really).

The rest is to decide the structs for input and output data.
Again, in the most primitive case we need just byte arrays. So long as we know input and output formats - we will be able to work with these.
But for the purpose of video (at least, although I may imagine a case when audio will need that too) these packages will also require "timestamps". These will help to sync video and audio pieces (in ffmpeg these all are called Frames).

As for the codec object itself, it needs to be able to register in the engine, possibly providing a list of extensions it would like to decode. Then engine will be able to try all registered plugin codecs, and then fall back it its own.

**Comment by ericoporto** [comment]
Here's an additional reference in a TheoraPlay Library from icculus (this mixes a bit with file/stream functionality, so not really what we want but it's small enough to think): https://github.com/icculus/DirkSimple/tree/main/thirdparty/theoraplay

The general idea I want is the minimal glue I need to make between arbitrary codecs and AGS video functionality.

> So for the very crude draft

This looks ok, I don't know enough to comment yet

> This kind of interface would work virtually for anything

ok, so you mentioned image. I have not thought about this for images, but I guess you are thinking about DynamicSprite.CreateFromFile and then we pass some format engine doesn't normally handle. This looks like a cool idea. 

I will say I think my biggest worry is on video because of the size of the video libraries vs the size of audio and image libraries - people are used with AGS being small and easy to build, and these video libraries are usually huge and use some esoteric building process.

I am good with having a more generic use though, if it can be done without adding too much complexity.

**Comment by ivan-mogilko** [comment]
You can make all three elements with similar API: encoder, decoder and filter. The difference between these is that decoder takes "raw data" (from file) and gives MediaData in the engine's format (or MediaFrame, if we use ffmpeg's terminology), encoder takes MediaData and gives "raw data" (something that may be saved back into file), and filter takes and gives MediaData, just different:

```
struct IMediaDecoder
{
     // we need to ask for an output format, which engine can work with
     result Open( output_media_format );
     result PutData(const uint8_t *raw_data, size_t length);
     result GetData(MediaData *out_data);
};

struct IMediaEncoder
{
     // we need to ask for an input format, which engine will pass into it;
     // but also maybe encoded format? question is how to pass it in a generic way
     result Open( input_media_format, encoded_format? );
     result PutData(const MediaData *in_data);
     result GetData(const uint8_t *raw_data, size_t *length);
};

struct IMediaFilter
{
     // we need to tell which format the engine is working in
     result Open(media_format);
     result PutData(const MediaData *in_data);
     result GetData(MediaData *out_data);
};
```

The big question about filter is: how to make sure it's applied where necessary, and not for every played clip.

About implementation in the engine: we need to make sure that:
1. every loading process has decoder split out from the rest.
2. our own "decoders" implement same interfaces as we expose to plugin API.

then every decoder becomes replaceable.

The SoundDecoder class is currently the one closest to this, I think. VideoPlayer is semi-close, but needs to be split into 2 parts (currently decoders are inheriting base "player" class instead).

**Comment by ivan-mogilko** [comment]
> Here's an additional reference in a TheoraPlay Library from icculus (this mixes a bit with file/stream functionality, so not really what we want but it's small enough to think): https://github.com/icculus/DirkSimple/tree/main/thirdparty/theoraplay

Yes, thinking more, and also looking at this, the above interface example is probably missing another "object" right before "decoder" (and right after "encoder"), because we need file streaming, and seeking too.

So, this might have to be merged with "Decoder/Encoder" in our case, for simplicity.

In which case, Decoder for example, does not have `PutData`, but accepts a "Stream" object in `Open`, and only `GetData` method. We'd need to think whether GetData is enough too, or it has to be separated into GetAudio/Video for convenience. OTOH it may be kept as GetData, and let decoder return whatever is ready in sync.

**Comment by ericoporto** [comment]
Hey, the last crude drafts look more approachable, but I am curious if it's possible to decode video and let the audio be figured out by the internal AGS audio decoder somehow, does this makes sense?

**Comment by ivan-mogilko** [comment]
> Hey, the last crude drafts look more approachable, but I am curious if it's possible to decode video and let the audio be figured out by the internal AGS audio decoder somehow, does this makes sense?

In this case we need the 2 step solution:
- Parse file and retrieve encoded packets, which we may then send into video and audio decoders;
- Decode packets and retrieve media data in the engine's format (simple wave + RGB bitmaps).

so there has to be 2 interfaces one for each such step; or 3 interfaces, that is - 1 for parsing file and splitting into video/audio packets, and +2 separate interfaces for video & audio decoding.

Encoding is the same, just in opposite order.

```
MediaFileReader ---- encoded audio packet  ------> AudioDecoder -- audio frame  -----> audio player
                            ---- encoded video packet  ------> VideoDecoder -- video frame ------> texture renderer
```

**Comment by ericoporto** [comment]
I was thinking if there would be a way to have built-in codecs and plugin codecs, in a way to force the code to be modularized regarding encoding. This [pl_mpeg](https://github.com/phoboslab/pl_mpeg) library for mpg-1 looks like a good candidate to being included and so that the code could be adjusted to support ogv and mpg-1 (with the idea to force thinking about how to select each encoder and how to separate things).

**Comment by ivan-mogilko** [comment]
Why do we need to support mpg-1, is there a benefit?

**Comment by ericoporto** [comment]
Ah right, I was not thinking to include it necessarily as in a way to support things, but just if we need simple code to experiment - I don't know how it would switch between one decoder and the other - this looks small enough and easy enough to build. Just something simple that would use the interface.
